### PR TITLE
Implement dht put and get CLI commands

### PR DESCRIPTION
**Pull request by whyrusleeping**
Implement `ipfs dht get` and `ipfs dht put`. Im open to some feedback on this, but this is just part one of a set of changes for getting ipns working well.

### CODE REVIEW

File: routing/dht/handlers.go [hunk 1]
```diff
@@ -83,7 +83,7 @@ func (dht *IpfsDHT) handleGetValue(ctx context.Context, p peer.ID, pmes *pb.Mess
 
 	// Find closest peer on given cluster to desired key and reply with that info
 	closer := dht.betterPeersToQuery(pmes, p, CloserPeerCount)
-	if closer != nil {
+	if len(closer) > 0 {
```
Comment (jbenet) [inline routing/dht/handlers.go]: curious why the change here? for readability or was there a bug?

File: routing/dht/dht.go [hunk 2]
```diff
@@ -322,7 +322,7 @@ func (dht *IpfsDHT) betterPeersToQuery(pmes *pb.Message, p peer.ID, count int) [
 	// == to self? thats bad
 	for _, p := range closer {
 		if p == dht.self {
-			log.Debug("Attempted to return self! this shouldnt happen...")
+			log.Error("Attempted to return self! this shouldnt happen...")
```
Comment (jbenet) [inline routing/dht/dht.go]: `log.Error` impacts the end user. we should no longer be using `log.Error`. use `log.Debug` or `log.Event`

File: routing/dht/routing.go [hunk 3]
```diff
@@ -84,7 +88,7 @@ func (dht *IpfsDHT) GetValue(ctx context.Context, key u.Key) ([]byte, error) {
 
 	// get closest peers in the routing table
 	rtp := dht.routingTable.ListPeers()
-	log.Debugf("peers in rt: %s", len(rtp), rtp)
+	log.Errorf("peers in rt: %s", len(rtp), rtp)
```
Comment (jbenet) [inline routing/dht/routing.go]: not `Error`.

File: routing/dht/routing.go [hunk 4]
```diff
@@ -153,7 +168,7 @@ func (dht *IpfsDHT) Provide(ctx context.Context, key u.Key) error {
 // FindProviders searches until the context expires.
 func (dht *IpfsDHT) FindProviders(ctx context.Context, key u.Key) ([]peer.PeerInfo, error) {
 	var providers []peer.PeerInfo
-	for p := range dht.FindProvidersAsync(ctx, key, math.MaxInt32) {
+	for p := range dht.FindProvidersAsync(ctx, key, KValue) {
```
Comment (jbenet) [inline routing/dht/routing.go]: why `KValue` and not max until it expires?  (just for allocation? is there another workaround?) will this break something else?

File: core/commands/dht.go [hunk 5]
```diff
@@ -352,3 +354,220 @@ var findPeerDhtCmd = &cmds.Command{
...
+					if verbose {
+						fmt.Fprintf(buf, "* querying %s\n", obj.ID)
+					}
+				case notif.Value:
+					fmt.Fprintf(buf, "got value: '%s'\n", obj.Extra)
+				case notif.QueryError:
+					fmt.Fprintf(buf, "error: %s\n", obj.Extra)
+				default:
+					fmt.Fprintf(buf, "unrecognized event type: %d\n", obj.Type)
+				}
```
Comment (jbenet) [inline core/commands/dht.go]: i think much of this printing out part can be common between commands.

**Comment by jbenet** [comment]
Sharness test pls. we shouldn't be adding more code without tests. I want to make fixing tests high priority. (it makes it annoying to develop.)

File: routing/dht/handlers.go (hunk 1 above)
Comment (whyrusleeping) [inline routing/dht/handlers.go]: mostly paranoid about this: http://play.golang.org/p/YSy0gaUefy

File: routing/dht/routing.go (hunk 3 above)
Comment (whyrusleeping) [inline routing/dht/routing.go]: whoops. my bad on that one.

File: routing/dht/dht.go (hunk 2 above)
Comment (whyrusleeping) [inline routing/dht/dht.go]: Hrm.. I feel like this should be an error case. But alright.

File: routing/dht/routing.go (hunk 4 above)
Comment (whyrusleeping) [inline routing/dht/routing.go]: Because its wasteful, we are getting to the point where there might actually be more than K providers of a value, and as it stands now, we will send out requests to each and every single one of them, which will make them send us blocks, lots of wasted bandwidth.

File: core/commands/dht.go (hunk 5 above)
Comment (whyrusleeping) [inline core/commands/dht.go]: me too... but the strings being printed are slightly different in each case. Ill take a look at refactoring.

**Comment by whyrusleeping** [comment]
@jbenet how should I write sharness tests for these? im not if i can bring up multiple nodes in a sharness test...

**Comment by whyrusleeping** [comment]
Also, would like to note that if it werent for the race tests running out of memory, we would have an entirely green test result
